Replace addPlugins with addSerializer and expose as method of expect

pedrottimark opened the pull request:
**Summary**

Fixes #2647

In `plugins.js` enclosed the body of arrow function in braces to keep from leaking the updated array. I assume API for new method of `expect` doesn’t intend to expose the plugins directly.

```js
exports.addSerializer = (plugin: any) => {
  PLUGINS = [plugin].concat(PLUGINS);
};
```

**Test plan**

* The existing tests for `foo` and `bar` in `integration_tests/snapshot_serializers` verify that replacing `addPlugins` with `addSerializer` works for `config.snapshotSerializers` of `package.json`
* Added 2 tests to `integration_tests/snapshot_serializers` in which `expect.addSnapshotSerializer` add
    * first serializer that overrides `foo` serializer in `config.snapshotSerializers` of `package.json`
    * second serializer that overrides both the first and `foo`
* Updated `plugins-test.js` in `jest-snapshot` and replaced one occurrence of arbitrary initial length also mutating `splice` with non-mutating `concat`
* Removed attempted mock and test in `jest-jasmine2` because as far as I could discover, calling `setup({config})` is not an effective simulation: the plugins array affected by `addSerializer` was different from the array returned by `getSerializers` so the only reason it passed in original test was `config.snapshotSerializers` always set to `[]` by mistake.

pedrottimark commented:
Oh yeah, because of work in progress on docs, I didn’t update them yet. Will write a first draft.

pedrottimark commented:
If there is a step to add or update snapshots in the `integration_tests/__tests__/__snapshots__` directory, then I missed it. Because the ordinary `npm test -- -u` doesn’t do it, I edited directly.

Do we need a step in CONTRIBUTING.md or somewhere else?

cpojer commented on packages/jest-jasmine2/src/__tests__/setup-jest-globals-test.js:
```diff
@@ -1,30 +0,0 @@
-  const setup = require('../setup-jest-globals');
-
-  beforeEach(() => {
-    require('jest-snapshot').__reset();
```
Can we also remove this __reset method? I don't think we need it any longer.

cpojer replied:
Oh nevermind, it was only part of the mock.

cpojer commented on packages/jest-matchers/src/index.js:
```diff
@@ -149,6 +152,8 @@ expect.objectContaining = objectContaining;
 expect.arrayContaining = arrayContaining;
 expect.stringMatching = stringMatching;
 
+expect.addSnapshotSerializer = addSerializer;
```
This will need to be added in the file that adds the globals in jest-jasmine2. Basically, jest-matchers cannot depend on jest-snapshot. jest-jasmine2 is the extension point :)

cpojer commented:
This looks good. Do you mind:

* Moving `expect.addSnapshotSerializers` to jest-jasmine2.
* Adding documentation for `expect.addSnapshotSerializers`? We published the new website and it is on master again, now, so it can be done in the same PR.

pedrottimark commented:
* Added a section following the asymmetric methods and preceding ordinary assertions.
* What you say about the extension point makes sense. However, I reverted my attempted change in the doc commit, because tests passed after I merged changes from master, but then the 2 new integration tests failed after I moved `expect.addSnapshotSerializer = addSerializer` to follow the following line:

https://github.com/facebook/jest/blob/master/packages/jest-jasmine2/src/jest-expect.js#L59

pedrottimark commented:
Here is a naive thought from inexperience with stateful modules. What if `jest-snapshot` exports a method that returns an object which encapsulates plugins by closure, similar to Redux store: https://github.com/facebook/jest/blob/master/packages/jest-jasmine2/src/jest-expect.js#L59

```js
() => ({
  let plugins = [ReactElementPlugin, ReactTestComponentPlugin];

  return {
    addSerializer: (plugin: any) => {
      plugins = [plugin].concat(plugins);
    },
    getSerializers: () => plugins,
  };
});
```

cpojer commented:
Nice work @pedrottimark

pedrottimark commented:
Are you sure it is okay to leave the extension point in jest-matchers, though?

cpojer commented:
Ohhh, no actually we have to fix that. I missed the comment I guess, I was too hasty.

pedrottimark commented:
Your thoughts on what is happening now are welcome. Why tests pass one way but not the other.

I was just now studying what it would look like if `plugins` was explicit part of `SnapshotState`. Include `snapshotSerializers` config as an arg to the constructor is as far as I got so far.

pedrottimark commented:
Here is another observation:

* `serialize` function could use `plugins` from `SnapshotState` at https://github.com/facebook/jest/blob/master/packages/jest-snapshot/src/State.js#L120
* if the function definition is moved inline from https://github.com/facebook/jest/blob/master/packages/jest-snapshot/src/utils.js#L61-L67
